### PR TITLE
⚠️ MHC: require control plane initialized, cluster infrastructure ready for node startup timeout

### LINKED ISSUE #3026: MachineHealthCheck issues handling machines right after the control plane becomes accessible
**What steps did you take and what happened:**
1. Create a MachineHealthCheck with some nodeStartupTimeout (e.g. 10m)
1. Create a cluster (Cluster, InfraCluster, KubeadmControlPlane, MachineDeployment, etc)
1. For whatever reason, the control plane takes a while to be accessible (close to nodeStartupTimeout, or more) - let's say it takes 9m
1. MachineHealthCheck evaluates unhealthiness based on the the machine's `.status.lastUpdated`, which only changes when we adjust `.status.phase`
1. Machine's status hasn't changed for a while (it can't get provisioned until the control plane is ready, as it's a worker node join)
1. MachineHealthCheck controller can finally talk to the workload cluster and logs something like this:

```
I0507 14:15:28.302500       1 machinehealthcheck_targets.go:232] controllers/MachineHealthCheck "msg"="Target is likely to go unhealthy" "Target"="default/w1/w1-md-0-6b5f6c66cf-kmrn6/" "cluster"="w1" "machinehealthcheck"="w1" "namespace"="default" "timeUntilUnhealthy"="1m41s"
```

In this case, it took ~ 8m19s for the control plane to be accessible, leaving only 1m41s for the machine to get a node ref...

**However**, because the machine's lastUpdated time changes as it changes phases (e.g. going from pending to provisioning because the control plane is now up), the deadline for getting a node ref is essentially extended each time lastUpdated changes.

**What did you expect to happen:**

1. Ideally, we don't include the time waiting for control plane accessibility in nodeStartupTimeout
1. We don't extend the node ref timeout each time the machine's phase changes

**Anything else you would like to add:**

**Environment:**

- Cluster-api version: v0.3.5
- Minikube/KIND version: v0.8.1
- Kubernetes version: (use `kubectl version`): v1.18.1

/kind bug
/area health

cc @JoelSpeed

### LINKED ISSUE #3798: Add Cluster ControlPlaneInitialized condition
**User Story**

As a developer, I would like to add a ControlPlaneInitialized condition to Clusters, so that MachineHealthCheck can use it when determining if a Machine has exceeded the allotted node startup timeout.

**Detailed Description**

We currently have cluster.status.controlPlaneInitialized, but it predates conditions and does not have a timestamp associated with it. Based on discussions in #3026, we want MachineHealthCheck to base nodeStartupTimeout calculations on the latter of (controlPlaneInitialized, machine creation timestamp).

This condition MUST be set to true as soon as we are able to establish connectivity with a workload cluster. We can keep the same logic we currently have for setting cluster.status.controlPlaneInitialized.

This condition MUST be a write-once value.

We also MUST remove cluster.status.controlPlaneInitialized from v1alpha4. As far as converting from 3 to 4, I'm not sure what we'd put in for LastTransitionTimestamp...

**Anything else you would like to add:**

#3779 is similar/related

/kind feature
/kind api-change
/milestone v0.4.0
/priority important-soon

### PR DESCRIPTION
**Pull request by ncdc**
**What this PR does / why we need it**:
Add ControlPlaneInitializedCondition.

Remove cluster.status.controlPlaneInitialized.

Change the MachineHealthCheck logic to require the Cluster's "control
plane initialized" and "infrastructure ready" conditions to be true
before proceeding to determine if a target is unhealhty.

We can't look just at a Machine's last updated time when determining if
the Machine has exceeded the node startup timeout.  A node can't
bootstrap until after the cluster's infrastructure is ready and the
control plane is initialized, and we need to base node startup timeout
on the latter of machine creation time, control plane initialized time,
and cluster infrastructure ready time.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3026
Fixes #3798

### CODE REVIEW

**Comment by vincepri** [comment]
/assign
/milestone v0.3.11

Going to review later today

File: controllers/machinehealthcheck_targets_test.go
Comment (ncdc) [inline controllers/machinehealthcheck_targets_test.go:36]: I've removed TestGetTargetsFromMHC as it's not really feasible to test in isolation with a live client (from envtest) when all the controllers are running (from suite_test). I have tried to move all the test cases from here into TestMachineHealthCheck_Reconcile.

File: controllers/machinehealthcheck_targets_test.go
Comment (ncdc) [inline controllers/machinehealthcheck_targets_test.go:311]: I've removed setting the client because healthCheckTargets() doesn't need one.

File: controllers/machinehealthcheck_targets.go
Comment (fabriziopandini) [inline controllers/machinehealthcheck_targets.go:120]: this should probably be ControlPlaneAvailable condition - when the first kubeadm API server is up and running -, assuming we are going to make this part of the contract (now AFAIK it exists only in KCP).

File: controllers/machinehealthcheck_targets.go
Comment (fabriziopandini) [inline controllers/machinehealthcheck_targets.go:123]: If I got the problem right, this is accurate only for the first controlplane machine, but for all the other machines the actual startup time triggers when status.ControlPlaneInitialized==True (or ControlPlaneAvailable condition = True as suggested above).
However, I'm wondering if this assumption is somehow related to how kubeadm works and can't be used as a generic rule...

File: controllers/machinehealthcheck_targets.go
Comment (ncdc) [reply controllers/machinehealthcheck_targets.go:120]: Yeah, that makes more sense. I'll change it. And we should definitely make it a control plane provider contract requirement in v1alpha4.

File: controllers/machinehealthcheck_targets.go
Comment (ncdc) [reply controllers/machinehealthcheck_targets.go:123]: Are you talking about KCP/MHC remediation?

In this PR's current form, given that MHC only remediates non-control plane Machines, this should be accurate for all non-control plane Machines.

Or am I misunderstanding what you're saying?

File: controllers/machinehealthcheck_controller_test.go
Comment (vincepri) [inline controllers/machinehealthcheck_controller_test.go:28]:
```suggestion
	. "github.com/onsi/gomega"
	
	capierrors "sigs.k8s.io/cluster-api/errors"
	"sigs.k8s.io/cluster-api/util/patch"
	corev1 "k8s.io/api/core/v1"
```

File: controllers/machinehealthcheck_controller_test.go
Comment (vincepri) [inline controllers/machinehealthcheck_controller_test.go:182]:
```suggestion
```

There should be a global context already defined, if there isn't you can definitely ignore this

File: controllers/machinehealthcheck_targets.go
Comment (vincepri) [reply controllers/machinehealthcheck_targets.go:123]: How is this going to impact the remediation support for control plane nodes?

**Review by vincepri — APPROVED** [review]
LGTM

File: controllers/machinehealthcheck_targets.go
Comment (ncdc) [reply controllers/machinehealthcheck_targets.go:123]: I need to take that into account, since when we switched to using conditions for MHC, we got rid of the logic that only has MHC work on Machines owned by MachineSets and never control plane machines.

**Comment by vincepri** [comment]
Changed the base branch to be release-0.3, we'll have to forward-port these changes to v0.4 later

**Comment by ncdc** [comment]
@vincepri any objections to resetting to main and I can backport to 0.3 after it's merged?

**Comment by vincepri** [comment]
Sure, up to you

File: controllers/machinehealthcheck_targets.go
Comment (detiber) [reply controllers/machinehealthcheck_targets.go:120]: Has it been documented/decided anywhere that controlPlaneRef will be a requirement? It seems like that will create a barrier to phased adoption of Cluster API, though I suppose one could create the equivalent of a null controlplane provider

File: controllers/machinehealthcheck_targets.go
Comment (detiber) [inline controllers/machinehealthcheck_targets.go:130]: How would this affect Machines that were provisioned long after the Cluster infrastructure is initialized, should this be updated to compare both the time since infraReadyTime and the machine last updated time against unhealthy duration to account for both cases? 

Alternatively, do we need a net new condition on a machine to indicate that provisioning has been started that can accurately be checked against rather than the approximations we are currently using?

File: controllers/machinehealthcheck_targets.go
Comment (detiber) [reply controllers/machinehealthcheck_targets.go:120]: Just to be clear, I'm not suggesting that we keep backward compat support for non-control plane provider managed control plane machines, just questioning the hard requirement on having a control plane provider implementation in place.

File: controllers/machinehealthcheck_targets.go
Comment (ncdc) [reply controllers/machinehealthcheck_targets.go:130]: Yeah I need to take that into account. This is on hold until I file a new issue about adding a condition for control plane available.

File: controllers/machinehealthcheck_targets.go
Comment (ncdc) [reply controllers/machinehealthcheck_targets.go:120]: @vincepri can you file an issue proposing controlPlaneRef as a requirement in v1alpha4

**Comment by ncdc** [comment]
I've updated this PR based on the latest comment in #3026. PTAL!

File: api/v1alpha3/conversion.go
Comment (ncdc) [inline api/v1alpha3/conversion.go:166]: Let me know if you'd prefer I do the ControlPlaneInitialized conversion in here. It's technically more accurate, but I wanted to use the `conditions.MarkTrue()` helper, and that requires a full root-level object (i.e. `Cluster`).

File: api/v1alpha3/conversion.go
Comment (vincepri) [inline api/v1alpha3/conversion.go:38]: Doesn't this mean that every time during a conversion we'd change the timestamp of the condition and cause the api server to fire off another patch event?

Comment (vincepri) [reply api/v1alpha3/conversion.go:38]: We should probably do this after we recover conditions and only set it if needed maybe?

File: api/v1alpha4/cluster_types.go
Comment (vincepri) [inline api/v1alpha4/cluster_types.go:133]: ⚠️ ?

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller_test.go
Comment (vincepri) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller_test.go:28]:
```suggestion
	"sigs.k8s.io/cluster-api/util/patch"
```

File: controllers/machinehealthcheck_targets.go
Comment (vincepri) [inline controllers/machinehealthcheck_targets.go:200]:
```suggestion
	if err := clusterClient.Get(ctx, client.ObjectKey{Namespace: mhc.Namespace, Name: mhc.Spec.ClusterName}, &cluster); err != nil {
```

File: api/v1alpha3/conversion.go
Comment (ncdc) [reply api/v1alpha3/conversion.go:38]: Doesn't `MarkTrue` & the underlying conditions helper code only change the timestamp if the status changes?

Comment (vincepri) [reply api/v1alpha3/conversion.go:38]: Ah, that's right

File: api/v1alpha4/cluster_types.go
Comment (ncdc) [reply api/v1alpha4/cluster_types.go:133]: Are we doing ⚠️ for anything & everything that is an API change from a3 to a4? If so, this PR is both 🐛 and ⚠️ - not sure what to do

Comment (vincepri) [reply api/v1alpha4/cluster_types.go:133]: Worth to get the ⚠️ in place, so when we get to release notes we can write down the impact in addition to the bug fix

**Comment by wfernandes** [comment]
The fix for the failing _flaky_ test  
```
I1218 21:25:31.856669   10410 machinehealthcheck_controller.go:105] controller/machinehealthcheck "msg"="Reconciling" "name"="test-mhc-lqvh6" "namespace"="test-mhc-rpvk7" "reconciler group"="cluster.x-k8s.io" "reconciler kind"="MachineHealthCheck" 
E1218 21:25:34.094583   10410 cluster_cache.go:361]  "msg"="Error health checking cluster" "error"="timed out waiting for the condition"  "cluster"="test-machine-watches-wd4dg/machine-reconcile-mfx5b"
    machinehealthcheck_controller_test.go:219: 
        Timed out after 5.002s.
        expected
        	&v1alpha4.MachineHealthCheckStatus{ExpectedMachines:2, CurrentHealthy:2, RemediationsAllowed:2, ObservedGeneration:1, Targets:[]string{"test-mhc-machine-9dvt4", "test-mhc-machine-ddpv6"}, Conditions:v1alpha4.Conditions{v1alpha4.Condition{Type:"RemediationAllowed", Status:"True", Severity:"", LastTransitionTime:v1.Time{Time:time.Time{wall:0x0, ext:63743923530, loc:(*time.Location)(0x2832fe0)}}, Reason:"", Message:""}}}
        to match
        	&v1alpha4.MachineHealthCheckStatus{ExpectedMachines:2, CurrentHealthy:2, RemediationsAllowed:2, ObservedGeneration:1, Targets:[]string{"test-mhc-machine-ddpv6", "test-mhc-machine-9dvt4"}, Conditions:v1alpha4.Conditions{v1alpha4.Condition{Type:"RemediationAllowed", Status:"True", Severity:"", LastTransitionTime:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, Reason:"", Message:""}}}
```

has actually been resolved in #3999 
I think we may have to rebase the machinehealthcheck_controller.go and its test.

**Comment by ncdc** [comment]
@wfernandes I am pretty sure my branch already included the commit from 3999. Maybe I accidentally dropped part of it when I was rebasing?

**Comment by ncdc** [comment]
Looks like some of the test cases are missing the sort - adding now.

**Comment by ncdc** [comment]
Odd failure:

```
[AfterSuite] AfterSuite
/home/prow/go/src/sigs.k8s.io/cluster-api/exp/controllers/suite_test.go:66
  Expected success, but got an error:
      <*errors.StatusError | 0xc00014c000>: {
          ErrStatus: {
              TypeMeta: {Kind: "", APIVersion: ""},
              ListMeta: {
                  SelfLink: "",
                  ResourceVersion: "",
                  Continue: "",
                  RemainingItemCount: nil,
              },
              Status: "Failure",
              Message: "Timeout: failed waiting for *v1alpha4.MachinePool Informer to sync",
              Reason: "Timeout",
              Details: {Name: "", Group: "", Kind: "", UID: "", Causes: nil, RetryAfterSeconds: 0},
              Code: 504,
          },
      }
      Timeout: failed waiting for *v1alpha4.MachinePool Informer to sync
  /home/prow/go/src/sigs.k8s.io/cluster-api/exp/controllers/suite_test.go:60
```

/test pull-cluster-api-test-main

**Comment by ncdc** [comment]
Looks like everything is passing now

**Comment by wfernandes** [comment]
Hmm...now it looks like we need to fix conflicts

**Comment by ncdc** [comment]
@wfernandes rebased

**Review by vincepri — APPROVED** [review]
/approve

**Comment by vincepri** [comment]
/hold 

for @fabriziopandini review

**Review by vincepri — APPROVED** [review]
/lgtm

**Comment by ncdc** [comment]
@fabriziopandini @vincepri 🤞 this should be ready to go

**Comment by ncdc** [comment]
@vincepri @fabriziopandini alright, looks like the tests are back to passing. PTAL again 😄

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**
